src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -646,18 +646,135 @@ lookAhead (Parser step1 initial1 _) = Parser step initial extract
 -------------------------------------------------------------------------------
 -- Interleaving
 -------------------------------------------------------------------------------
+
+data ParserTurn = Parser1 | Parser2
+
 --
 -- | See 'Streamly.Internal.Data.Parser.deintercalate'.
 --
--- /Unimplemented/
+-- /Internal/
 --
 {-# INLINE deintercalate #-}
 deintercalate ::
-    -- Monad m =>
+    MonadCatch m =>
        Fold m a y -> Parser m x a
     -> Fold m b z -> Parser m x b
     -> Parser m x (y, z)
-deintercalate = undefined
+deintercalate
+    (Fold fstep1 finitial1 fextract1)
+    (Parser pstep1 pinitial1 pextract1)
+    (Fold fstep2 finitial2 fextract2)
+    (Parser pstep2 pinitial2 pextract2) =
+
+    Parser step initial extract
+
+    where
+
+    initial = do
+        finit1 <- finitial1
+        pinit1 <- pinitial1
+        finit2 <- finitial2
+        pinit2 <- pinitial2
+        return (finit1, pinit1, finit2, pinit2, Parser1, 0::Int)
+
+    step (fs1, ps1, fs2, ps2, currentParser, numBuffered) a =
+        case currentParser of
+            Parser1 -> do
+                st <- pstep1 ps1 a
+                case st of
+                    Partial n ps1new ->
+                        return $ Partial n (fs1, ps1new, fs2, ps2,
+                            currentParser, 0)
+                    Continue n ps1new -> do
+                        let
+                            newNumBuffered =
+                                if n==0
+                                then numBuffered + 1
+                                else numBuffered - n
+                        return $ Continue n (fs1, ps1new, fs2, ps2,
+                            currentParser, newNumBuffered)
+                    Done n result -> do
+                        res1 <- fstep1 fs1 result
+                        pinit1 <- pinitial1
+                        pinit2 <- pinitial2
+                        case res1 of
+                            FL.Partial fs1new ->
+                                return $ Partial n (fs1new, pinit1, fs2, pinit2,
+                                    Parser2, numBuffered)
+                            FL.Done result1 -> do
+                                result2 <- fextract2 fs2
+                                return $ Done numBuffered (result1, result2)
+                    Error _ -> do
+                        -- input definitively rejected by the first parser
+                        pinit1 <- pinitial1
+                        pinit2 <- pinitial2
+                        return $ Partial numBuffered (fs1, pinit1, fs2, pinit2,
```

Comment:
You should backtrack `numBuffered + 1` elements here?
And restart the buffer count from `0` instead of `numBuffered - 1`?